src/event.c

```diff
@@ -410,6 +636,153 @@ void module_event(const char *module, const char *event, struct ua *ua,
 }
 
 
+static int event_emit(struct event event, enum ua_event ev, const char *fmt,
+		      va_list ap)
+{
+	struct le *le;
+	char buf[256];
+	int n;
+
+	n = re_vsnprintf(buf, sizeof(buf), fmt, ap);
+	if (n == -1)
+		return EINVAL;
+
+	event.ev  = ev;
+	event.txt = buf;
+	event.err = 0;
+	le = ehel.head;
+	while (le) {
+		struct ehe *ehe = le->data;
+		le = le->next;
+
+		if (ehe->h(&event, ehe->arg))
+			break;
+	}
+
+	if (event.err)
+		return event.err;
+
+	/* backwards compatibility */
+	ua_event(event.u.ua, ev, event.u.call, event.txt);
+
+	return 0;
+}
+
+
+/**
+ * Emit an application event
+ *
+ * @param ev   User-agent event
+ * @param fmt  Formatted arguments
+ * @param ...  Variable arguments
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int event_app_emit(enum ua_event ev, const char *fmt, ...)
+{
+	va_list ap;
+	struct event event = {.ev = ev};
+	int err;
+
+	if (event_class(ev) != EVENT_CLASS_APP)
+		return EINVAL;
+
+	va_start(ap, fmt);
+	err = event_emit(event, ev, fmt, ap);
+	va_end(ap);
+
+	return err;
+}
+
+
+/**
+ * Emit a User-Agent event
+ *
+ * @param ev   User-Agent event
+ * @param ua   User-Agent
+ * @param fmt  Formatted arguments
+ * @param ...  Variable arguments
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int event_ua_emit(enum ua_event ev, struct ua *ua, const char *fmt, ...)
+{
+	struct event event = {.ev = ev};
+	va_list ap;
+	int err;
+
+	if (event_class(ev) != EVENT_CLASS_UA)
+		return EINVAL;
+
+	event.u.ua = ua;
+
+	va_start(ap, fmt);
+	err = event_emit(event, ev, fmt, ap);
+	va_end(ap);
+
+	return err;
+}
+
+
+/**
+ * Emit a Call event
+ *
+ * @param ev    User-Agent event
+ * @param call  Call object
+ * @param fmt   Formatted arguments
+ * @param ...   Variable arguments
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int event_call_emit(enum ua_event ev, struct call *call, const char *fmt, ...)
+{
+	struct event event = {.ev = ev};
+	va_list ap;
+	int err;
+
+	if (event_class(ev) != EVENT_CLASS_CALL)
+		return EINVAL;
+
+	event.u.call = call;
+
+	va_start(ap, fmt);
+	err = event_emit(event, ev, fmt, ap);
+	va_end(ap);
+
+	return err;
+}
+
+
+/**
+ * Emit a SIP message event
+ *
+ * @param ev    User-Agent event
+ * @param msg   SIP message
+ * @param fmt   Formatted arguments
+ * @param ...   Variable arguments
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int event_sip_msg_emit(enum ua_event ev, const struct sip_msg *msg,
+		       const char *fmt, ...)
+{
+	struct event event = {.ev = ev};
+	va_list ap;
+	int err;
+
+	if (event_class(ev) != EVENT_CLASS_CALL)
```

Comment:
Is `EVENT_CLASS_CALL` right here?

Reply:
Good point!
```c
	if (event_class(ev) != EVENT_CLASS_SIP)
		return EINVAL;
```
is correct.

Explanation:
The event class selects the union field.

| event class | union field |
| --- | --- |
| EVENT_CLASS_UA | event.u.ua |
| EVENT_CLASS_CALL | event.u.call |
| EVENT_CLASS_APP | event.nothing |
| EVENT_CLASS_SIP  |  event.u.msg |

Problem
---
The function `ua_event_class_name()` returns some incongruous class names
- `register`, better would be `ua`
- `mwi` --> better `ua`
- `UA_EVENT_CREATE` returns `application` --> better `ua`
- `VU_REPORT` --> better `call`

But maybe there are already applications which event parsing rely on the current texts. Thus I didn't touch this function. The function is used in `event_encode_dict()` which I would mark as deprecated.

The new function `odict_encode_event()` uses `event_class_name()` that returns logical fitting class names. This should also make parsing of events simpler for applications.